akka-typed/src/main/scala/akka/typed/cluster/Cluster.scala

```diff
@@ -0,0 +1,314 @@
+ * subscribers) followed by [[MemberStatus]] `Exiting` and finally [[MemberStatus]] `Removed`.
+ *
+ * Note that this command can be issued to any member in the cluster, not necessarily the
+ * one that is leaving. The cluster extension, but not the actor system or JVM, of the
+ * leaving member will be shutdown after the leader has changed status of the member to
+ * Exiting. Thereafter the member will be removed from the cluster. Normally this is
+ * handled automatically, but in case of network failures during this process it might
+ * still be necessary to set the node’s status to Down in order to complete the removal.
+ */
+final case class Leave(address: Address) extends ClusterCommand
+
+/**
+ * Send command to DOWN the node specified by 'address'.
+ *
+ * When a member is considered by the failure detector to be unreachable the leader is not
+ * allowed to perform its duties, such as changing status of new joining members to 'Up'.
+ * The status of the unreachable member must be changed to 'Down', which can be done with
+ * this method.
+ */
+final case class Down(address: Address) extends ClusterCommand
+
+/**
+ * Akka Typed Cluster API entry point
+ */
+object Cluster extends ExtensionId[Cluster] {
+
+  def createExtension(system: ActorSystem[_]): Cluster = new AdapterClusterImpl(system)
+
+}
+
+/**
+ * INTERNAL API:
+ */
+@InternalApi
+private[akka] object AdapterClusterImpl {
+
+  private def subscriptionsBehavior(adaptedCluster: akka.cluster.Cluster) = Actor.deferred[ClusterStateSubscription] { ctx ⇒
+    var selfSeenUp = false
+    var upSubscribers: List[ActorRef[SelfUp]] = Nil
+    var removedSubscribers: List[ActorRef[SelfRemoved.type]] = Nil
+
+    adaptedCluster.subscribe(ctx.self.toUntyped, ClusterEvent.initialStateAsEvents, classOf[MemberEvent])
+
+    // important to not eagerly refer to it or we get a cycle here
+    def cluster = Cluster(ctx.system)
+    def updateCachedSelfMember(member: Member): Unit = {
+      // safe because we are the only writer
+      cluster.asInstanceOf[AdapterClusterImpl]._cachedSelfMember = member
+    }
+
+    Actor.immutable[AnyRef] { (ctx, msg) ⇒
+
+      msg match {
+        case Subscribe(subscriber, eventClass) ⇒
+          adaptedCluster.subscribe(subscriber.toUntyped, initialStateMode = ClusterEvent.initialStateAsEvents, eventClass)
+          Actor.same
+
+        case Unsubscribe(subscriber) ⇒
+          adaptedCluster.unsubscribe(subscriber.toUntyped)
+          Actor.same
+
+        case OnSelfUp(subscriber) ⇒
+          if (cluster.selfMember.status == MemberStatus.Up) subscriber ! SelfUp(adaptedCluster.state)
+          else if (!selfSeenUp) {
+            ctx.watch(subscriber)
+            upSubscribers = subscriber :: upSubscribers
+          } else {
+            // self did join, but is now no longer up, we want to avoid subscribing
+            // to not get a memory leak, but also not signal anything
```

Comment:
instead of checking `cluster.selfMember.status` you could have a flag `ignoreSelfUpSubscription` that is set when you get MemberLeft, MemberExited or MemberRemoved

instead of the flags it should probably be an enum (sealed trait states), something like
BeforeUp, Up, AfterUp, Removed